### PR TITLE
168736413 controller dbcontext

### PR DESCRIPTION
**Pull request by uchan3**
# [add the address controller + dbContext], [#168736413]

## reviewer

- @fredbelotte

## contributing

this project is maintained under the terms of the [Contribution Guidelines][contribution-guidelines-url].

[contribution-guidelines-url]: https://github.com/revaturexyz/addressxyz/blob/master/.github/CONTRIBUTING.md 'CONTRIBUTION GUIDELINES'

### CODE REVIEW

File: .docker/docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# addressxyz :: docker-compose
+
+networks:
+  addressxyz_localhost:
+    driver: bridge
```
Comment (fredbelotte) [inline .docker/docker-compose.yaml]: remove this directive...

File: .docker/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  api:
+    depends_on:
+      - "db"
+    deploy:
+      replicas: 1
+    image: addressxyz
+    networks:
+      - addressxyz_localhost
+    ports:
+      #Map 10080 port as Guest port and 80 port as Host port on vmbox ->
```
Comment (fredbelotte) [inline .docker/docker-compose.yaml]: remove these comments

File: aspnet/Xyz.Address.Api/Controllers/AddressController.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            return Ok(addressModel);   
+        }
+        
+        ///<summary> Endpoint to add a address, which is validated using Google Maps Geocoding API </summary>
+        ///<param name="newAddress"> The model that is created from the Query String </param>
+        ///<return> if(success) returns a status code 200 Response (else) returns a status code 404 response. </return>
+        [HttpPost]
+        [Route("postaddress")]
+        [Filters.ValidateAddressWithGoogle]
+        //http://localhost:5000/api/address/postaddress?street=156315+Smoketree+St
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Controllers/AddressController.cs]: remove this line

File: aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace Xyz.Address.Api.Filters{
+  public class ValidateAddressWithGoogleAttribute: ActionFilterAttribute{
+    public override void OnActionExecuting(ActionExecutingContext context){
+      var rawQueryString = context.HttpContext.Request.QueryString.ToString();
+      rawQueryString = rawQueryString.Substring(1); //removing the ? at the front
+      var addressModelProperties = rawQueryString.Split("&");
+      context.Result = new JsonResult(addressModelProperties);
+      return ;
+      // Domain.AddressModel _model = new Domain.AddressModel();
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs]: remove all of these lines

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 5]
```diff
@@ -26,8 +27,35 @@ public Startup(IConfiguration configuration)
...
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
-            services.AddDbContext<AddressDbContext>();
+
+            ///<summary>
+            /// We are fetching the connection string for postgres from appsetting.json.
+            ///</summary>
+            var ConnectionString = Configuration["ConnectionStrings"];
+
+            ///<summary>
+            /// Adding AddressDbContext as a service.
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: summary is too long

File: aspnet/Xyz.Address.Lib/Models/AddressModel.cs [hunk 6]
```diff
@@ -21,44 +20,55 @@ public AddressModel()
...
     ///<param name="ValidationContext"> List of Address Properties to be validated on. </param>
     ///<return> If property is null, the Validate method returns an error message, indicating property cannot be null. </return>
     public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
     {
-      foreach (var property in this.GetType().GetProperties())
-      {
-        if (property.GetValue(property) is null)
+      foreach (var Property in this.GetType().GetProperties()) 
+      { 
+        //System.Reflection.TargetException: Object does not match the target type.
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Lib/Models/AddressModel.cs]: remove these lines

File: aspnet/Xyz.Address.Tests/Address.Api.Tests.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        public void TestAddressControllerGet()
+        {
+          var AC = new AddressController();
+          var result = AC.Get();
+
+         // Assert.IsType<OkObjectResult>(result);
+                   Assert.IsType<NotFoundObjectResult>(result);
+        }
+
+        // TODO FIX
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Tests/Address.Api.Tests.cs]: remove these lines

File: .docker/dockerup.yaml [hunk 8]
```diff
@@ -6,21 +6,28 @@ networks:
...
       replicas: 1
     image: postgres
+    ports:
+      - "5432:5432"
+    restart: always
     networks:
       - addressxyz_localhost
+    environment:
+      POSTGRES_USER: "postgres"
+      POSTGRES_PASSWORD: ""
```
Comment (fredbelotte) [inline .docker/dockerup.yaml]: give it a password

File: aspnet/Xyz.Address.Api/Controllers/AddressController.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            if (addressModel.Street.Length == 0) return NotFound(string.Format("Key {0}, is invalid,", Key));
+            return Ok(addressModel);   
+        }
+        
+        ///<summary> Endpoint to add a address, which is validated using Google Maps Geocoding API </summary>
+        ///<param name="newAddress"> The model that is created from the Query String </param>
+        ///<return> if(success) returns a status code 200 Response (else) returns a status code 404 response. </return>
+        [HttpPost]
+        [Route("postaddress")]
+        // [Filters.ValidateAddressWithGoogle]
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Controllers/AddressController.cs]: remove this line

File: aspnet/Xyz.Address.Api/Controllers/AddressController.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+          }
+          return NotFound("Could not add the given address");
+        }
+
+        ///<summary> Endpoint to update a address, which is validated using Google Maps Geocoding API </summary>
+        ///<param name="newAddress"> The model that is created from the Query String </param>
+        ///<return> if(success) returns a status code 200 Response (else) returns a status code 404 response. </return>
+        [HttpPut]
+        [Route("putaddress")]
+        //[Filters.ValidateAddressWithGoogle]
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Controllers/AddressController.cs]: remove this line

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 11]
```diff
@@ -26,12 +28,38 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            // Add Global exception handling class
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: remove this line

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 12]
```diff
@@ -26,12 +28,38 @@ public Startup(IConfiguration configuration)
...
+            services.AddMvc(options =>
+            {
+              options.Filters.Add(new GlobalExceptionHandlerAttribute());
+            });
+
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
-            services.AddDbContext<AddressDbContext>();
+
+            ///<summary>
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: remove this... xml comments are only for class-level or method-level

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 13]
```diff
@@ -26,12 +28,38 @@ public Startup(IConfiguration configuration)
...
+            /// We are fetching the connection string for postgres from appsetting.json.
+            ///</summary>
+            var ConnectionString = Configuration["ConnectionStrings"];
+
+            ///<summary>
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: remove all of these lines

File: aspnet/Xyz.Address.Api/appsettings.json [hunk 14]
```diff
@@ -4,5 +4,6 @@
       "Default": "Warning"
     }
   },
-  "AllowedHosts": "*"
+  "AllowedHosts": "*",
+  "ConnectionStrings" : "Host=db;port=54321;Database=postgres;Username=postgres;Password="
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/appsettings.json]: set password, change port to 5432

File: aspnet/Xyz.Address.Dbx/AddressDbContext.cs [hunk 15]
```diff
@@ -1,15 +1,47 @@
...
+  /// The DbContext is responsible for opening and managing connections to the database. It is the primary class responsible for 
+  /// interacting with data as objects DbContext.
+  /// </summary>
   public class AddressDbContext : DbContext
   {
-    public DbSet<AddressEntity> Addresses;
+    public DbSet<AddressEntity> Addresses {get; set;}
+    public AddressDbContext(){}
+
+    //NEED
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/AddressDbContext.cs]: what is this?

File: aspnet/Xyz.Address.Dbx/AddressDbContext.cs [hunk 16]
```diff
@@ -1,15 +1,47 @@
...
+    public AddressDbContext(DbContextOptions<AddressDbContext> options): base(options){
+    }
+
+    protected override void OnConfiguring(DbContextOptionsBuilder builder){
+      builder.UseNpgsql("Host=db;port=5432;Database=postgres;Username=postgres;Password=");
+    }
+
+    /// <summary>
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/AddressDbContext.cs]: too long of a summary

File: aspnet/Xyz.Address.Dbx/Entities/AddressEntity.cs [hunk 17]
```diff
@@ -18,22 +18,8 @@ public AddressEntity()
...
-        Street = this.Street,
-        City = this.City,
-        Country = this.Country,
-        PostalCode = this.PostalCode,
-        StateProvince = this.StateProvince
-      };
-    }
-    #endregion
-
+    ///<summary>Active: describes if Address is active or no
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/Entities/AddressEntity.cs]: remove this

File: aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs [hunk 18]
```diff
@@ -3,31 +3,37 @@
...
-///</summary>
+// using Microsoft.EntityFrameworkCore.Migrations;
+// using Microsoft.EntityFrameworkCore;
 
 namespace Xyz.Address.Dbx.Repositories
 {
+    ///<summary> Create concrete address entity for Data Repository Pattern. Implements CRUD Operations defined in IRepository. </summary>
     public class AddressRepository: IRepository
     {
+      ///<summary>Dependency injection of AddressDbContext into AddressRepository</summary>
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs]: remove this line

File: aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs [hunk 19]
```diff
@@ -3,31 +3,37 @@
...
       private AddressDbContext _db;
 
       #region Constructor
       public AddressRepository(AddressDbContext context){
         _db = context;
+        // _db.Database.Migrate();
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs]: use or lose this line

File: aspnet/Xyz.Address.Lib/Models/AddressModel.cs [hunk 20]
```diff
@@ -21,44 +20,51 @@ public AddressModel()
 
     #region Methods
 
-    ///<summary> Perform validation checks on the Address Properties.  </summary>
+    ///<summary> Perform null validation checks on the Address Properties.  </summary>
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Lib/Models/AddressModel.cs]: remove this thing

File: aspnet/Xyz.Address.Tests/ValidAddress.cs [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        Street = "1560 Broadway St.",
+        City = "Colorado", //Note that City and StateProvince are not equal which is what we want to test.
+        StateProvince = "Colorado", 
+        PostalCode = "80112",
+        Country = "USA"
+      };
+
+      //Act
+      var validationResult = new List<ValidationResult>();
+      //var actual = Validator.TryValidateProperty(validAddress, new ValidationContext(validAddress), validationResult);
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Tests/ValidAddress.cs]: remove this line

File: aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    public override void OnActionExecuting(ActionExecutingContext context){
+      var rawQueryString = context.HttpContext.Request.QueryString.ToString();
+      rawQueryString = rawQueryString.Substring(1); //removing the ? at the front
+      var addressModelProperties = rawQueryString.Split("&");
+      context.Result = new JsonResult(addressModelProperties);
+      return ;
+    }
+
+    public async Task<string> CallingGoogleApi(string street, string city, string state, HttpClient http){
+      //Google Geocoding API
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs]: remove this line

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 23]
```diff
@@ -26,12 +28,24 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: remove this line

File: aspnet/Xyz.Address.Api/Startup.cs [hunk 24]
```diff
@@ -26,12 +28,24 @@ public Startup(IConfiguration configuration)
...
+            var ConnectionString = Configuration["ConnectionStrings"];
+
+            services.AddEntityFrameworkNpgsql().AddDbContext<AddressDbContext>(options => options.UseNpgsql(ConnectionString, npgsql =>{
+              npgsql.EnableRetryOnFailure(); //we will continue to attempt to connect to the database
+            }));             
+          
+            services.AddScoped<AddressDbContext>();
         }
 
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Api/Startup.cs]: remove this line

File: aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs [hunk 25]
```diff
@@ -3,12 +3,12 @@
 using System.Linq;
 using Xyz.Address.Dbx.Models;
 using Domain = Xyz.Address.Lib.Models;
-// ADD XML COMMENTS
-///<summary>Create concrete address entity for Data Repository Pattern. Implements CRUD Operations defined in IRepository.
-///</summary>
+// using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/Repositories/AddressRepository.cs]: remove these lines

File: aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs (hunk 22 above)
Comment (sudeepcb) [reply aspnet/Xyz.Address.Api/Filters/ValidateAddressWithGoogle.cs]: done

File: aspnet/Xyz.Address.Api/Startup.cs (hunk 23 above)
Comment (sudeepcb) [reply aspnet/Xyz.Address.Api/Startup.cs]: done

File: aspnet/Xyz.Address.Dbx/AddressDbContext.cs (hunk 15 above)
Comment (fredbelotte) [reply aspnet/Xyz.Address.Dbx/AddressDbContext.cs]: clean this up

File: aspnet/Xyz.Address.Dbx/AddressDbContext.cs [hunk 26]
```diff
@@ -1,15 +1,42 @@
...
+  /// The DbContext is responsible for opening and managing connections to the database. It is the primary class responsible for 
+  /// interacting with data as objects DbContext.
+  /// </summary>
   public class AddressDbContext : DbContext
   {
-    public DbSet<AddressEntity> Addresses;
+    public DbSet<AddressEntity> Addresses {get; set;}
+    public AddressDbContext(){}
+
+    //NEED
```
Comment (fredbelotte) [inline aspnet/Xyz.Address.Dbx/AddressDbContext.cs]: what is this????

**Pull request merged**